src/bitops.c

```diff
@@ -779,34 +779,53 @@ void bitopCommand(client *c) {
     addReplyLongLong(c,maxlen); /* Return the output string length in bytes. */
 }
 
-/* BITCOUNT key [start end] */
+/* BITCOUNT key [start end [BIT|BYTE]] */
 void bitcountCommand(client *c) {
     robj *o;
-    long start, end, strlen;
+    long long start, end, strlen;
+    long tmplen;
     unsigned char *p;
     char llbuf[LONG_STR_SIZE];
+    int isbit = 0, prefix = 0, postfix = 0;
 
     /* Lookup, check for type, and return 0 for non existing keys. */
     if ((o = lookupKeyReadOrReply(c,c->argv[1],shared.czero)) == NULL ||
         checkType(c,o,OBJ_STRING)) return;
-    p = getObjectReadOnlyString(o,&strlen,llbuf);
+    p = getObjectReadOnlyString(o,&tmplen,llbuf);
+    strlen = tmplen;
 
     /* Parse start/end range if any. */
-    if (c->argc == 4) {
-        if (getLongFromObjectOrReply(c,c->argv[2],&start,NULL) != C_OK)
+    if (c->argc == 4 || c->argc == 5) {
+        if (getLongLongFromObjectOrReply(c,c->argv[2],&start,NULL) != C_OK)
             return;
-        if (getLongFromObjectOrReply(c,c->argv[3],&end,NULL) != C_OK)
+        if (getLongLongFromObjectOrReply(c,c->argv[3],&end,NULL) != C_OK)
             return;
         /* Convert negative indexes */
         if (start < 0 && end < 0 && start > end) {
             addReply(c,shared.czero);
             return;
         }
+        if (c->argc == 5) {
+            if (!strcasecmp(c->argv[4]->ptr,"bit")) isbit = 1;
+            else if (!strcasecmp(c->argv[4]->ptr,"byte")) isbit = 0;
+            else {
+                addReplyErrorObject(c,shared.syntaxerr);
+                return;
+            }
+        }
+        if (isbit) strlen <<= 3;
         if (start < 0) start = strlen+start;
         if (end < 0) end = strlen+end;
         if (start < 0) start = 0;
         if (end < 0) end = 0;
         if (end >= strlen) end = strlen-1;
+        if (isbit && start <= end) {
+            strlen >>= 3;
+            prefix = start&7;
+            postfix = 7-(end&7);
```

Comment:
i don't like the names `prefix` and `postfix`, maybe something like this will be clearer and also simplify the code below?
```suggestion
            /* Before converting bit offset to byte offset, create negative masks for the edges. */
            first_byte_neg_mask = (1<<(start&7))-1;
            last_byte_neg_mask = (~(1<<(end&7))-1) & 0x0ff;
```